openwisp_controller/config/tasks.py

```diff
@@ -114,6 +211,150 @@ def trigger_vpn_server_endpoint(endpoint, auth_token, vpn_id):
+        vpn.ip.ip_address,
+        instance=vpn,
+        action='update_member',
+        info=(
+            f'Successfully updated ZeroTier network member: {vpn.node_id}, '
+            f'ZeroTier network: {vpn.network_id}, '
+            f'ZeroTier VPN server UUID: {vpn_id}'
+        ),
+        err=(
+            f'Failed to update ZeroTier network member: {vpn.node_id}, '
+            f'ZeroTier network: {vpn.network_id}, '
+            f'ZeroTier VPN server UUID: {vpn_id}'
+        ),
+    )
+
+
+@shared_task(
+    bind=True,
+    base=OpenwispApiTask,
+    autoretry_for=(RequestException,),
+    **API_TASK_RETRY_OPTIONS,
+)
+def trigger_zerotier_server_join(self, vpn_id):
+    Vpn = load_model('config', 'Vpn')
+    vpn = Vpn.objects.get(pk=vpn_id)
+    service_method = ZerotierService(
+        vpn.host,
+        vpn.auth_token,
+    ).join_network
+    response = self.handle_api_call(
+        service_method,
+        vpn.network_id,
+        instance=vpn,
+        action='network_join',
+        info=(
+            f'Successfully joined the ZeroTier network: {vpn.network_id}, '
+            f'ZeroTier VPN Server UUID: {vpn_id}'
+        ),
+        err=(
+            f'Failed to join ZeroTier network: {vpn.network_id}, '
+            f'VPN Server UUID: {vpn_id}'
+        ),
+    )
+    if response.ok:
+        # Update zerotier network controller
+        trigger_zerotier_server_update_member.delay(vpn_id)
+
+
+@shared_task(
+    bind=True,
+    base=OpenwispApiTask,
+    autoretry_for=(RequestException,),
+    **API_TASK_RETRY_OPTIONS,
+)
+def trigger_zerotier_server_delete(self, host, auth_token, network_id, vpn_id):
+    service_method = ZerotierService(host, auth_token).delete_network
+    response = self.handle_api_call(
+        service_method,
+        network_id,
+        info=(
+            f'Successfully deleted the ZeroTier VPN Server '
+            f'with UUID: {vpn_id}, Network ID: {network_id}'
+        ),
+        err=(
+            'Failed to delete ZeroTier VPN Server '
+            f'with UUID: {vpn_id}, Network ID: {network_id}, '
+            'as it does not exist on the ZeroTier Controller Networks'
+        ),
+        send_notification=False,
+    )
+    # In case of successful deletion of the network
+    # we should also remove controller node from the network
+    if response.status_code == 200:
+        trigger_zerotier_server_leave.delay(host, auth_token, network_id)
+
+
+@shared_task(
+    bind=True,
+    base=OpenwispApiTask,
+    autoretry_for=(RequestException,),
+    **API_TASK_RETRY_OPTIONS,
+)
+def trigger_zerotier_server_leave(self, host, auth_token, network_id):
+    service_method = ZerotierService(host, auth_token).leave_network
+    self.handle_api_call(
+        service_method,
+        network_id,
+        info=f'Successfully left the ZeroTier Network with ID: {network_id}',
+        err=f'Failed to leave ZeroTier Network with ID: {network_id}',
+        send_notification=False,
+    )
+
+
```

Comment:
I think it's time to start separating tasks like these, what about creating a file called tasks_zerotier.py and importing the tasks here? @pandafy @Aryamanz29

Reply:
I believe that would be better :+1: